source/agora/consensus/validation/PreImage.d

```diff
@@ -47,21 +51,39 @@ version (unittest)
 *******************************************************************************/
 
 public string isInvalidReason (const ref PreImageInfo new_image,
-    const ref PreImageInfo prev_image, uint validator_cycle) nothrow @safe
+    const ref PreImageInfo prev_image, uint validator_cycle, Hash random_seed)
+    nothrow @safe
 {
-    if (new_image.enroll_key != prev_image.enroll_key)
-        return "The pre-image's enrollment key differs from its descendant";
-
-    if (new_image.distance < prev_image.distance)
-        return "The height of new pre-image is smaller than that of previous one";
-
     if (new_image.distance > validator_cycle)
         return "The hashing count of two pre-images is above the validator cycle";
 
+    ulong distance;
+    Hash prev_hash;
+    if (prev_image != PreImageInfo.init)
+    {
+        if (new_image.enroll_key != prev_image.enroll_key)
+            return "The pre-image's enrollment key differs from its descendant";
+
+        if (new_image.distance < prev_image.distance)
+            return "The height of new pre-image is smaller than that of previous one";
+
+        distance = new_image.distance - prev_image.distance;
+        prev_hash = prev_image.hash;
+    }
+    else if (random_seed != Hash.init)
+    {
+        distance = new_image.distance + 1;
+        prev_hash = random_seed;
+    }
+    else
+    {
+        return "The random seed is invalid";
+    }
+
```

Comment:
We discussed this IRL, but I think it's important to re-state it: We should not change the algorithm to adapt to the data. Or at least, not this much. We're essentially duplicating a part of the validation routine to adapt for the special case of the `random_seed`. Except that it's not a special case. One can just think of the `random_seed` as the *first* pre-image, and store it in the table. This way, we always have a pre-image, and we don't need all this complication.